### ISSUE TITLE
FP implicit-clone: &&Path.to_path_buf()

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#8227]
### Summary

clippy suggestes invalid code when cloning `&&Path`s 

### Lint Name

clippy::implicit-clone

### Reproducer

I tried this code:

```rust
use std::path::PathBuf;

fn main() {
    let a: Vec<&PathBuf> = Vec::new();
    let mut c: Vec<PathBuf> = Vec::new();

    for path /* &&PathBuf */  in a.iter() {
        c.push(path.to_path_buf());
    }
}

```

I saw this happen:
`cargo clippy -- -Wclippy::implicit_clone `
```
    Checking clippy_bug v0.1.0 (/tmp/clippy_bug)
warning: implicitly cloning a `PathBuf` by calling `to_path_buf` on its dereferenced type
 --> src/main.rs:8:21
  |
8 |         c.push(path.to_path_buf());
  |                     ^^^^^^^^^^^ help: consider using: `clone`
  |
  = note: requested on the command line with `-W clippy::implicit-clone`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implicit_clone
```

I expected to see this happen:

I expected the suggested code to compile, but it didn't :sweat_smile: 

````
error[E0308]: mismatched types
 --> src/main.rs:8:16
  |
8 |         c.push(path.clone());
  |                ^^^^^^^^^^^^- help: try using a conversion method: `.to_path_buf()`
  |                |
  |                expected struct `std::path::PathBuf`, found `&std::path::PathBuf`

````
Maybe clippy does not understand that we have a double-reference here?

### Version

```text
clippy 0.1.59 (7d6f948 2022-01-04)

rustc 1.59.0-nightly (7d6f94817 2022-01-04)
binary: rustc
commit-hash: 7d6f948173ccb18822bab13d548c65632db5f0aa
commit-date: 2022-01-04
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
